src/lib/project-saver-hoc.jsx

```diff
@@ -118,25 +119,36 @@ const ProjectSaverHOC = function (WrappedComponent) {
          * @param {?object} requestParams - object of params to add to request body
          */
         storeProject (projectId, requestParams) {
-            return this.props.vm.saveProjectSb3()
-                .then(content => {
-                    const assetType = storage.AssetType.Project;
-                    const dataFormat = storage.DataFormat.SB3;
-                    const body = new FormData();
-                    body.append('sb3_file', content, 'sb3_file');
-                    if (requestParams) {
-                        for (const key in requestParams) {
-                            body.append(key, requestParams[key]);
-                        }
-                    }
-                    // when id is undefined or null, storage.store as we have
-                    // configured it will create a new project with id
-                    return storage.store(
-                        assetType,
-                        dataFormat,
-                        body,
-                        projectId
-                    );
+            requestParams = requestParams || {};
+            return Promise.all(this.props.vm.assets
+                .filter(asset => !asset.clean)
+                .map(
+                    asset => storage.store(
+                        asset.assetType,
+                        asset.dataFormat,
+                        asset.data,
+                        asset.assetId
+                    ).then(
+                        () => (asset.clean = true)
+                    )
+                )
+            ).then(() => {
+                const body = new FormData();
+                const sb3Json = new Blob([this.props.vm.toJSON()], {type: 'application/json'});
+                body.append('sb3_file', sb3Json, 'sb3_file');
+                for (const key in requestParams) {
+                    if (requestParams.hasOwnProperty(key)) body.append(key, requestParams[key]);
```

Comment:
I vaguely understand that using hasOwnProperty like this is the recommended pattern. But I don't understand why. There are no other properties that actually get iterated in plain old `for (const key in requestParams) {...}`, right? Is there an example of code I can run where I can see this problem occur?

Reply:
Good question!

Here's some code:

```javascript
class Parent { }

const alice = new Parent()

// inheritted property is also enumerable
Parent.prototype.greeting = "Hi."

for(key in alice) {
	console.log("Enumerable: " + key) // "greeting"
	console.log("Own property? " + alice.hasOwnProperty(key)) // "false"
}
```

Most of the default properties which come with JavaScript are not enumerable by default. However, if you make new ones, they might appear in `for...in` loops. Such loops detect keys which are higher up in the prototype chain, though.

Reply:
It's pretty bizarre, I have no recollection of writing this! I meant to leave any existing code unchanged. I'm really at a loss of how this got here 😕 

It's generally safer to use this check whenever you're using `for ... in` for the reason @joker314 said — you only want to enumerate the values you've personally added to the object. But it doesn't matter in this case.